src/python/py/models/builder.py

```diff
@@ -1792,7 +1950,7 @@ def make_layer(self, layer_id, layer):
             # Norm after last decoder layer of model (last layer --> norm)
             self.layernorm_attrs["last_layernorm"] = True
 
-    def make_model(self, input_path):
+    def make_model(self, input_path, adapter_path):
```

Comment:
We should remove `adapter_path` because the `input_path` should contain all necessary files (PyTorch weights, JSON files, etc). Otherwise, we would have to provide separate arguments for the paths to many types of files and that is not a scalable approach.

Reply:
adapter_config.json and adapter safetensors file will be in separate folder right? Also when we are using a quantized model we have to pass input_path so it will be conflict between adapter path and input path?

Reply:
@kunal-vaishnavi akshay and I had a long discussion about this. 
- The adapter config saves the path/name of whatever the model it was trained on. this means that if an adapter is shared on hf-hub or through some other means the base_model_name_or_path might not be valid anymore.
  - valid case: 
    - was trained on an hf-id model like `meta-llama/Llama-2-7b-hf`
    - trained on a local model "local_path" and ran builder from the same place
  - invalid case:
   - trained on a non hf-id model like a locally download model, azureml registry model, github model and model builder run from a different machine
   
From my point of view, it is safer and easier logic wise to treat the base model and adapters separately. We don't know how or where the adapters were trained on so we cannot reliably get the base model information from the adapter config. This also limits the support for local adapters + the valid case above. 
For example, it would not work for an adapter on the model hub like https://huggingface.co/Mikael110/llama-2-7b-guanaco-qlora/blob/main/adapter_config.json

Reply:
tldr: adapter path is not a special case of the model format like gguf is

Reply:
For PyTorch models, the model builder only works for models that are compatible with Hugging Face's APIs currently. For example, it will fail if you provide `meta-llama/Llama-2-7b` as the model name and it will pass if you provide `meta-llama/Llama-2-7b-hf` as the model name.

Since we assume that the base PyTorch model can be loaded with `AutoModelForCausalLM.from_pretrained()`, I think it is fine to assume that the adapters can be loaded with `AutoPeftModelForCausalLM.from_pretrained()` since it comes from Hugging Face's `peft` package and `AutoPeftModelForCausalLM.from_pretrained()` will anyways invoke `AutoModelForCausalLM.from_pretrained()` when loading the base model. This also lets `AutoPeftModelForCausalLM.from_pretrained()` handle the local path vs. model name differences.

Most Hugging Face repos have the correct `base_model_name_or_path` entry. For the ones that don't, one option is that we can recommend users to download the repo locally, fix the `base_model_name_or_path` entry in the `adapter_config.json` file, and then run the model builder.

Alternatively, we can use the existing `-m/--model_name` and `-i/--input_path` flags to [load a `PeftModel`](https://github.com/huggingface/peft/blob/749b92456218f7dddc8f7a9aa27a41815b3d6c2e/src/peft/peft_model.py#L427) in the [same way that a GGUF model is loaded](https://github.com/microsoft/onnxruntime-genai/blob/main/src/python/py/models/README.md#gguf-model).

```python
from peft import PeftModel

peft_model = PeftModel.from_pretrained(args.model_name, args.input_path)
```

By using both the `-m` and `-`i flags, the `-a/--adapter_path` is not needed.

> For example, it would not work for an adapter on the model hub like https://huggingface.co/Mikael110/llama-2-7b-guanaco-qlora/blob/main/adapter_config.json

Yes, this would anyways fail since the repo is not compatible with the Hugging Face APIs.

Reply:
this is not a blocker for the olive-based workflow but just wanted to share the different scenarios for the adapters. 

https://huggingface.co/Mikael110/llama-2-7b-guanaco-qlora/blob/main/adapter_config.json was actually trained on https://huggingface.co/meta-llama/Llama-2-7b-hf (as per their readme and my local usage). the `"/root/llama2-7b-hf"` was the local path where the base model was downloaded before the finetuning.

Maybe the alternative is a good compromise. My main concern is that coupling the base model and adapter model limits the use case scenarios. perhaps it's acceptable trade-off for you.
Another thing I want to note is that the base model can also be already int4 quantized using awq or gptq + adapters in float. In that case, `AutoPeftModelForCausalLM.from_pretrained()` probably won't work since the builder loads quantized models differently.

Reply:
just a heads up, this is also a use case:
- quantized base model, float adapters

what would happen in that case? both adapter and quantized model use `i`/`input_path`

Reply:
It looks to me that adapter_config.json is adding more special cares. Adding an extra argument is simpler and less confused here. Is there any other concern other than adding new argument?

Reply:
I added the check for `adapter_config.json` while I was adding the LoRA support because adapters created by Hugging Face's `peft` have an associating `adapter_config.json` file. Since we have agreed to add the `adapter_path` in the `extra_options` for now and revisit this design later, I think we can remove the check for `adapter_config.json`.

Reply:
Yes, I will make the changes